File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int answer = -1;
    
    public void check(int k, int[][] dungeons, List<Boolean> visited, int count) {
        if (count > answer) answer = count;
        for (int i = 0; i < visited.size(); i++) {
            if (!visited.get(i) && dungeons[i][0] <= k) {
                visited.set(i, true);
                check(k - dungeons[i][1], dungeons, visited, count + 1);
                visited.set(i, false);
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        List<Boolean> visited = new ArrayList<>();
        for (int i = 0; i < dungeons.length; i++) {
            visited.add(false);
        }
        
        check(k, dungeons, visited, 0);
        return answer;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    
    def check(k, dungeons, visited, count):
        nonlocal answer
        if count > answer:
            answer = count
            
        for i in range(len(dungeons)):
            if not visited[i] and dungeons[i][0] <= k:
                visited[i] = True
                check(k - dungeons[i][1], dungeons, visited, count + 1)
                visited[i] = False
                
    visited = [False] * len(dungeons)
    check(k, dungeons, visited, 0)
    
    return answer